go-controller/pkg/ovn/libovsdbops/meter.go

```diff
@@ -0,0 +1,71 @@
+package libovsdbops
+
+import (
+	"context"
+	"fmt"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+)
+
+// FindMeterByName finds an ovn meter by name, will return `errNotFound` if it does not exist
+func FindMeterByName(nbClient libovsdbclient.Client, name string) (*nbdb.Meter, error) {
+	searched := &nbdb.Meter{
+		Name: name,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), types.OVSDBTimeout)
+	defer cancel()
+	err := nbClient.Get(ctx, searched)
+	if err != nil {
+		return nil, err
+	}
+
+	return searched, nil
+}
+
+// UpdateMeterFairness updates the `fair` column of an ovn Meter, should only be called if a
+// copy of the meter has already been acquired with FindMeter
```

Comment:
The comment is a bit confusing as the only thing we are using from `meter` is the name which is the same you use with `FindMeterByName`. So you don't really need to have called `FindMeterByName` previously. This method could have `name string` as input instead of `meter *nbdb.Meter`

Reply:
So this was used in a funny scenario... In the master code we need to either create or Update a Meter + accompanying meterBand.  However currently there is no safe way to lookup a meter band (no name or identifiable fields) .... Meaning we will always end up doing multiple transactions  since with something like this 
i.e
1. A Meter Lookup 
IF IT EXISTS 
2. Update Meter Fairness 
IF IT DOES NOT EXIST 
2. Create a Meter-Band and a Meter in a single transactions 

There's no way to take care of both permutations in a single transaction because 
```
opModels := []OperationModel{
		{
			Model: meterBand,
			DoAfter: func() {
				meter.Bands = []string{meterBand.UUID}
			},
		},
		{
			Model: meter,
		},
	}

m := NewModelClient(nbClient)
	if _, err := m.CreateOrUpdate(opModels...); err != nil {
		return fmt.Errorf("error while creating Meter_Band %+v and Meter %+v error %v", meterBand, meter, err)
	}
```

Would ALWAYS try and create a new Meter Band even if the meter+meter band already existed

Reply:
So short answer we might as well just perform an update rather than a mutate on the meter (in the case it already existed) since we already had to look it up/ put a copy of it in memory

Reply:
Unless something like this is allowed I guess 
```
tmpMeter := &sbdb.Meter{
Name: meterName
Fair: &true, 
}
existingMeters := *[]sbdb.Meter{}
existingMeterBand := *[]sbdb.MeterBand{}

tmpMeterBand := &sbdb.MeterBand{}

opModel := []OperationModel{
{		
	Model: tmpMeter
	OnModelMutations: []interface{}{
		&tmpMeter.Fair
	},
	ExistingResult: existingMeters,
	DoAfter: func() {
		if existingMeters == 1 {
			// Now meter band wont be created if it exists 
			tmpMeterBand = existingMeters[0].Bands[0]
		}
	},
}, 
{
	Model: tmpMeterBand
	ExistingResult: existingMeterBand,
	DoAfter: func() {
		if existingMeterBand == 1 {
			tmpMeter.Bands = existingMeterBand[0].UUID
		} else {
			tmpMeter.Bands = tmpMeterBand.UUID
		}



	},
}, 
}


		{
			Model: meterBand,
			DoAfter: func() {
				meter.Bands = []string{meterBand.UUID}
			},
		},
		{
			Model: meter,
		},
	}

Reply:
What's easier to read in your opinion?

Reply:
I was just saying consider changing this
```
// UpdateMeterFairness updates the `fair` column of an ovn Meter, should only be called if a
// full in memory version of the meter has already been acquired with FindMeter
func UpdateMeterFairness(nbClient libovsdbclient.Client, meter *nbdb.Meter, fairness bool) error {
```
to this
```
// UpdateMeterFairness updates the `fair` column of an ovn Meter.
func UpdateMeterFairness(nbClient libovsdbclient.Client, meterName string, fairness bool) error {
```
No strong opinions anyway.

But I like you final suggestion because it would take car of updating the meter ban if required with `Action: types.MeterAction` and `Rate: config.Logging.ACLLoggingRateLimit` (at least the second option seems like something we might update).